# Question 3

1. In this part, you are required to implement Apriori and FPgrowth algorithms on the data provided as "data_apriori_fpgrowth/data.csv". The dataset is of online directory of certified businesses.

2. We have loaded the dataset for you.

3. You are free to use any library or implement your own algorithms.


**Use minimum support : 0.1**

In [16]:
import pandas as pd
import numpy as np
from mlxtend.frequent_patterns import apriori, fpgrowth
from mlxtend.preprocessing import TransactionEncoder

In [21]:
data = np.load("data_apriori_fpgrowth/data.npy", allow_pickle = True)
te = TransactionEncoder()
te_ary = te.fit(data).transform(data)
df = pd.DataFrame(te_ary, columns=te.columns_)
len(te.columns_)

721

# To Do
1. Run Apriori Algorithm on the dataset using minimum support 0.1.
 1. Note the run time of the algorithm to find frequent item sets.
 1. Print all the frequent item sets found by apriori.
 1. Print all frequent item sets of length $2$ or higher with support of $0.2$ or higher. 
1. Run FPGrowth Algorithm on the dataset using minimum support 0.1.
 1. Note the run time of the FPGrowth algorithm to find frequent item sets.
 1. Print all frequent item sets found by FPGrowth.
 1. Print all frequent item sets of length $2$ or higher with support of $0.2$ or higher.
1. Answer the following question:
 1. Both algorithms find and return exactly the same frequent itemsets (given same parameters of support and itemset length). What difference do you note in both of the algorithms?
 1. There were more than $1000$ transactions. Why there are too few itemsets returned by the algorithms?
 

# Apriori [10 marks]

In [10]:
# Apriori
frequent_items = apriori(df, min_support=0.1, use_colnames=True)

In [11]:
# All Frequent Itemsets
frequent_items

support                       itemsets
0   0.202113                        (ASIAN)
1   0.300704                        (BLACK)
2   0.152113                     (Brooklyn)
3   0.164085                     (HISPANIC)
4   0.671127                          (MBE)
5   0.300000                 (NON-MINORITY)
6   0.295070                     (New York)
7   0.477465                          (WBE)
8   0.200000                   (MBE, ASIAN)
9   0.300704                   (BLACK, MBE)
10  0.112676                (MBE, Brooklyn)
11  0.164085                (HISPANIC, MBE)
12  0.170423                (MBE, New York)
13  0.169014                     (MBE, WBE)
14  0.118310       (NON-MINORITY, New York)
15  0.300000            (NON-MINORITY, WBE)
16  0.175352                (WBE, New York)
17  0.118310  (NON-MINORITY, WBE, New York)

In [14]:
# Frequent itemsets of length 2 with support of 0.2
frequent_items[frequent_items.apply(lambda x: x['support'] > 0.2 and len(x['itemsets']) ==2, axis=1)]

support             itemsets
9   0.300704         (BLACK, MBE)
15  0.300000  (NON-MINORITY, WBE)

# FPGrowth [10 marks]

In [17]:
# FPGrowth
frequent_items = fpgrowth(df, min_support=0.1, use_colnames=True)

In [18]:
# All Frequent Itemsets
frequent_items

support                       itemsets
0   0.671127                          (MBE)
1   0.477465                          (WBE)
2   0.300704                        (BLACK)
3   0.152113                     (Brooklyn)
4   0.202113                        (ASIAN)
5   0.295070                     (New York)
6   0.164085                     (HISPANIC)
7   0.300000                 (NON-MINORITY)
8   0.169014                     (MBE, WBE)
9   0.300704                   (BLACK, MBE)
10  0.112676                (MBE, Brooklyn)
11  0.200000                   (MBE, ASIAN)
12  0.175352                (WBE, New York)
13  0.170423                (MBE, New York)
14  0.118310       (NON-MINORITY, New York)
15  0.118310  (NON-MINORITY, WBE, New York)
16  0.164085                (HISPANIC, MBE)
17  0.300000            (NON-MINORITY, WBE)

In [19]:
# Frequent itemsets of length 2 with support of 0.2
frequent_items[frequent_items.apply(lambda x: x['support'] > 0.2 and len(x['itemsets']) ==2, axis=1)]

support             itemsets
9   0.300704         (BLACK, MBE)
17  0.300000  (NON-MINORITY, WBE)

# Answer to the Questions [5 marks]

<p> With these inputs there is just a very slight difference in speed of the two algorithms but as number of transactions increases we can expect more and more difference in speed</p>

<p>There was a lot of variation in possible items in the transactions. In total there were 721 different items and just above 1000 transactions and average items in transactions were also few so that's why there were only a few itemsets returned which high support</p>